test/config-luatest/credentials_applier_test.lua

```diff
@@ -72,23 +72,6 @@ local function define_assert_priv_function(server)
 end
 
 g.test_converters = function()
```

Comment:
Nit: I would name it 'test_converter_from_config` or somehow similar, because now it verifies only one converter.

We can add a separate `test_converter_from_box`, which runs a server, actually grants privileges in the config origin and calls `privileges_from_box`. OTOH, it is called in other test cases and it seems it is covered to some extent anyway. I wouldn't push you to work on the new test cases for the box converter. It would be a right thing, but it may cost some time that we don't have now.

Reply:
How about such diff?

<details><summary>diff</summary>

```diff
diff --git a/test/config-luatest/credentials_applier_test.lua b/test/config-luatest/credentials_applier_test.lua
index 343403077..ef02b3946 100644
--- a/test/config-luatest/credentials_applier_test.lua
+++ b/test/config-luatest/credentials_applier_test.lua
@@ -72,251 +72,168 @@ local function define_assert_priv_function(server)
 end
 
 g.test_converters = function()
-    -- Guest privileges in format provided by config schema
-    local config_guest_data = {
-        privileges = {{
-                permissions = {
-                    'session',
-                    'usage'
-                },
-                universe = true,
-            }, {
-                permissions = {
-                    'execute',
-                },
-                lua_eval = true,
-                lua_call = {
-                    'all'
-                },
-            }
-        },
+    local credentials = {
         roles = {
-            'public'
-        },
-    }
-
-    -- Guest privileges in format of intermediate representation
-    local intermediate_guest_privileges = {
-        ['user'] = {},
-        ['role'] = {
-            ['public'] = {
-                ['execute'] = true
-            }
-        },
-        ['space'] = {},
-        ['function'] = {},
-        ['sequence'] = {},
-        ['universe'] = {
-            [''] = {
-                ['session'] = true,
-                ['usage'] = true,
-            }
-        },
-        ['lua_eval'] = {
-            [''] = {
-                ['execute'] = true,
-            },
-        },
-        ['lua_call'] = {
-            [''] = {
-                ['execute'] = true,
-            },
-        },
-        ['sql'] = {},
-    }
-
-    t.assert_equals(internal.privileges_from_config(config_guest_data),
-                    intermediate_guest_privileges)
-
-
-    local config_admin_data = {
-        privileges = {{
-                permissions = {
-                    'read',
-                    'write',
-                    'execute',
-                    'session',
-                    'usage',
-                    'create',
-                    'drop',
-                    'alter',
-                    'reference',
-                    'trigger',
-                    'insert',
-                    'update',
-                    'delete',
+            myrole = {
+                roles = {'super'},
+                privileges = {
+                    {
+                        permissions = {'read'},
+                        spaces = {'_priv'},
+                    },
+                    {
+                        permissions = {'execute'},
+                        functions = {'box.schema.user.info'},
+                    },
                 },
-                universe = true,
             },
-        }
-    }
-
-    local intermediate_admin_privileges = {
-        ['user'] = {},
-        ['role'] = {},
-        ['space'] = {},
-        ['function'] = {},
-        ['sequence'] = {},
-        ['universe'] = {
-            [''] = {
-                ['read'] = true,
-                ['write'] = true,
-                ['execute'] = true,
-                ['session'] = true,
-                ['usage'] = true,
-                ['create'] = true,
-                ['drop'] = true,
-                ['alter'] = true,
-                ['reference'] = true,
-                ['trigger'] = true,
-                ['insert'] = true,
-                ['update'] = true,
-                ['delete'] = true,
-            }
         },
-        ['lua_eval'] = {},
-        ['lua_call'] = {},
-        ['sql'] = {},
-    }
-
-    t.assert_equals(internal.privileges_from_config(config_admin_data),
-                    intermediate_admin_privileges)
-
-    local config_replication_data = {
-        privileges = {{
-                permissions = {
-                    'write'
-                },
-                spaces = {
-                    '_cluster',
-                },
-            }, {
-                permissions = {
-                    'read'
+        users = {
+            myuser = {
+                roles = {'myrole', 'replication'},
+                privileges = {
+                    {
+                        permissions = {'read', 'drop'},
+                        spaces = {'_space', '_index'},
+                    },
+                    {
+                        permissions = {'write', 'create'},
+                        spaces = {'_func', '_space'},
+                    },
+                    {
+                        permissions = {'write'},
+                        sequences = {''},
+                    },
+                    {
+                        permissions = {'execute'},
+                        functions = {'LUA'},
+                        sql = {'all'},
+                        lua_call = {'all'},
+                        lua_eval = true,
+                    },
                 },
-                universe = true,
-            },
-        }
-    }
-
-    local intermediate_replication_privileges = {
-        ['user'] = {},
-        ['role'] = {},
-        ['space'] = {
-            ['_cluster'] = {
-                ['write'] = true,
             },
-        },
-        ['function'] = {},
-        ['sequence'] = {},
-        ['universe'] = {
-            [''] = {
-                ['read'] = true,
+            guest = {
+                roles = {'super'},
             },
         },
-        ['lua_eval'] = {},
-        ['lua_call'] = {},
-        ['sql'] = {},
     }
 
-    t.assert_equals(internal.privileges_from_config(config_replication_data),
-                    intermediate_replication_privileges)
-
-    local config_custom_data = {
-        privileges = {{
-                permissions = {
-                    'read',
-                    'write',
+    local function verify()
+        local config = require('config')
+        local credentials = require('internal.config.applier.credentials')
+        local internal = credentials._internal
+        local box_privs = internal.privileges_from_box
+        local cfg_privs = internal.privileges_from_config
+
+        -- Intermediate representation of myrole privileges.
+        local exp = {
+            ['user'] = {},
+            ['role'] = {
+                super = {
+                    execute = true,
                 },
-                spaces = {
-                    'all',
+            },
+            ['space'] = {
+                _priv = {
+                    read = true,
                 },
-                sequences = {
-                    'myseq1',
-                    'myseq2',
+            },
+            ['function'] = {
+                ['box.schema.user.info'] = {
+                    execute = true,
+                }
+            },
+            ['sequence'] = {},
+            ['universe'] = {},
+            ['lua_eval'] = {},
+            ['lua_call'] = {},
+            ['sql'] = {},
+        }
+        t.assert_equals(box_privs('myrole'), exp)
+        t.assert_equals(cfg_privs(config:get('credentials.roles.myrole')), exp)
+
+        -- Intermediate representation of guest privileges.
+        exp = {
+            ['user'] = {},
+            ['role'] = {
+                super = {
+                    execute = true,
                 },
-            }, {
-                permissions = {
-                    'execute',
+            },
+            ['space'] = {},
+            ['function'] = {},
+            ['sequence'] = {},
+            ['universe'] = {},
+            ['lua_eval'] = {},
+            ['lua_call'] = {},
+            ['sql'] = {},
+        }
+        t.assert_equals(box_privs('guest'), exp)
+        t.assert_equals(cfg_privs(config:get('credentials.users.guest')), exp)
+
+        -- Intermediate representation of myuser privileges.
+        exp = {
+            ['user'] = {},
+            ['role'] = {
+                myrole = {
+                    execute = true,
                 },
-                functions = {
-                    'myfunc1',
-                    'myfunc2',
+                replication = {
+                    execute = true,
                 },
-            }, {
-                permissions = {
-                    'read',
+            },
+            ['space'] = {
+                _func = {
+                    create = true,
+                    write = true,
                 },
-                universe = true,
-            }, {
-                permissions = {
-                    'execute',
+                _index = {
+                    drop = true,
+                    read = true,
                 },
-                lua_call = {
-                    'all'
+                _space = {
+                    create = true,
+                    write = true,
+                    drop = true,
+                    read = true,
                 },
             },
-        },
-        roles = {
-            'myrole1',
-            'myrole2',
-            'public',
-        }
-    }
-
-    local intermediate_custom_privileges = {
-        ['user'] = {},
-        ['role'] = {
-            ['myrole1'] = {
-                ['execute'] = true,
-            },
-            ['myrole2'] = {
-                ['execute'] = true,
-            },
-            ['public'] = {
-                ['execute'] = true,
-            },
-        },
-        ['space'] = {
-            [''] = {
-                ['read'] = true,
-                ['write'] = true,
-            },
-        },
-        ['function'] = {
-            ['myfunc1'] = {
-                ['execute'] = true,
-            },
-            ['myfunc2'] = {
-                ['execute'] = true,
+            ['function'] = {
+                LUA = {
+                    execute = true,
+                },
             },
-        },
-        ['sequence'] = {
-            ['myseq1'] = {
-                ['read'] = true,
-                ['write'] = true,
+            ['sequence'] = {
+                [''] = {
+                    write = true,
+                },
             },
-            ['myseq2'] = {
-                ['read'] = true,
-                ['write'] = true,
+            ['universe'] = {},
+            ['lua_eval'] = {
+                [''] = {
+                    execute = true,
+                },
             },
-        },
-        ['universe'] = {
-            [''] = {
-                ['read'] = true,
+            ['lua_call'] = {
+                [''] = {
+                    execute = true,
+                },
             },
-        },
-        ['lua_eval'] = {},
-        ['lua_call'] = {
-            [''] = {
-                ['execute'] = true,
+            ['sql'] = {
+                [''] = {
+                    execute = true,
+                },
             },
-        },
-        ['sql'] = {},
-    }
+        }
+        t.assert_equals(box_privs('myuser'), exp)
+        t.assert_equals(cfg_privs(config:get('credentials.users.myuser')), exp)
+    end
 
-    t.assert_equals(internal.privileges_from_config(config_custom_data),
-                    intermediate_custom_privileges)
+    helpers.success_case(g, {
+        options = {credentials = credentials},
+        verify = verify,
+    })
 end
 
 g.test_privileges_subtract = function()
```

</details>

Reply:
I'm OK.